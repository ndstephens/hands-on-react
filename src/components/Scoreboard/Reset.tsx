import styled from '@emotion/styled';

import { useMouseDown } from '@/hooks/useMouseDown';

const BUTTON_ICON = 'ðŸ˜¬';
const BUTTON_ICON_MOUSE_DOWN = 'ðŸ˜²';

interface Props {
  /**
   * Reset action handler
   */
  onReset: () => void;
}

export const Reset = ({ onReset }: Props) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  return (
    <Button
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onMouseLeave={onMouseUp}
      onClick={onReset}
      data-testid="reset-button"
    >
      {mouseDown ? BUTTON_ICON_MOUSE_DOWN : BUTTON_ICON}
    </Button>
  );
};

const Button = styled.button`
  cursor: pointer;
  font-size: 1.5vw;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;

  &:active {
    border-color: #9e9e9e white white #9e9e9e;
  }
`;
