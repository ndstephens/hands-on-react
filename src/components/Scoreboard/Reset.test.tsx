import { fireEvent, render, screen } from '@testing-library/react';

import { Reset } from './Reset';

const BUTTON_ICON = 'ðŸ˜¬';
const BUTTON_ICON_MOUSE_DOWN = 'ðŸ˜²';

describe('Reset button test', () => {
  // Default component instance
  const ResetWithDummyHandler = () => <Reset onReset={() => null} />;

  it('Should render correctly', () => {
    const { asFragment } = render(<ResetWithDummyHandler />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render elements with default state', () => {
    render(<ResetWithDummyHandler />);

    expect(screen.getByTestId('reset-button')).toBeInTheDocument();
  });

  it('onReset handler should be called', () => {
    const onReset = jest.fn();
    render(<Reset onReset={onReset} />);
    fireEvent.click(screen.getByTestId('reset-button'));

    expect(onReset).toBeCalledTimes(1);
  });

  it('Should change state when onMouseDown and onMouseUp event happen', () => {
    render(<ResetWithDummyHandler />);

    const btn = screen.getByTestId('reset-button');

    fireEvent.mouseDown(btn);
    expect(btn.innerHTML).toBe(BUTTON_ICON_MOUSE_DOWN);
    fireEvent.mouseUp(btn);
    expect(btn.innerHTML).toBe(BUTTON_ICON);
  });

  it('Should change state when onMouseDown and onMouseLeave events happen', () => {
    render(<ResetWithDummyHandler />);

    fireEvent.mouseDown(screen.getByTestId('reset-button'));
    expect(screen.getByTestId('reset-button').innerHTML).toBe(
      BUTTON_ICON_MOUSE_DOWN
    );
    fireEvent.mouseLeave(screen.getByTestId('reset-button'));
    expect(screen.getByTestId('reset-button').innerHTML).toBe(BUTTON_ICON);
  });
});
