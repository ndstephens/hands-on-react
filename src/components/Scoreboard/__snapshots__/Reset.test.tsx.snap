// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset button test Should render correctly 1`] = `
<DocumentFragment>
  <button
    class="css-1cjjgsg"
    data-testid="reset-button"
  >
    ðŸ˜¬
  </button>
</DocumentFragment>
`;
